src/collectors/prometheus_rabbitmq_core_metrics_collector.erl

```diff
@@ -117,50 +117,43 @@
+        {2, io_sync_ops_total, counter, "Total number of I/O sync operations", io_sync_count},
+        {2, io_sync_time_microseconds_total, counter, "Total I/O sync time", io_sync_time},
+        {2, io_seek_ops_total, counter, "Total number of I/O seek operations", io_seek_count},
+        {2, io_seek_time_microseconds_total, counter, "Total I/O seek time", io_seek_time},
+        {2, io_open_attempt_ops_total, counter, "Total number of file open attempts", io_file_handle_open_attempt_count},
+        {2, io_open_attempt_time_microseconds_total, counter, "Total file open attempts time", io_file_handle_open_attempt_time},
+        {2, io_reopen_ops_total, counter, "Total number of times files have been reopened", io_reopen_count},
+        {2, schema_db_ram_tx_total, counter, "Total number of Schema DB memory transactions", mnesia_ram_tx_count},
+        {2, schema_db_disk_tx_total, counter, "Total number of Schema DB disk transactions", mnesia_disk_tx_count},
+        {2, msg_store_read_total, counter, "Total number of Message Store read operations", msg_store_read_count},
+        {2, msg_store_write_total, counter, "Total number of Message Store write operations", msg_store_write_count},
+        {2, queue_index_read_ops_total, counter, "Total number of Queue Index read operations", queue_index_read_count},
+        {2, queue_index_write_ops_total, counter, "Total number of Queue Index write operations", queue_index_write_count},
+        {2, queue_index_journal_write_ops_total, counter, "Total number of Queue Index Journal write operations", queue_index_journal_write_count}
     ]},
 
     {ra_metrics, [
-        {2, raft_term, counter, "Raft member term"},
-        {3, raft_log_snapshot_index, counter, "Raft log snapshot index"},
-        {4, raft_log_last_applied_index, counter, "Raft log last applied index"},
-        {5, raft_log_commit_index, counter, "Raft log commit index"},
-        {6, raft_log_last_written_index, counter, "Raft log last written index"},
+        {2, raft_term, gauge, "Raft member term"},
+        {3, raft_log_snapshot_index, gauge, "Raft log snapshot index"},
+        {4, raft_log_last_applied_index, gauge, "Raft log last applied index"},
+        {5, raft_log_commit_index, gauge, "Raft log commit index"},
+        {6, raft_log_last_written_index, gauge, "Raft log last written index"},
         {7, raft_entry_commit_latency, gauge, "Time taken for an entry to be committed"}
     ]},
 
     {queue_coarse_metrics, [
         {2, queue_messages_ready, gauge, "Messages ready to be delivered to consumers"},
         {3, queue_messages_unacked, gauge, "Messages delivered to consumers but not yet acknowledged"},
         {4, queue_messages, gauge, "Sum of ready and unacknowledged messages - total queue depth"},
-        {5, erlang_process_reductions, counter, "Erlang process reductions"}
+        {5, erlang_process_reductions_total, counter, "Total number of Erlang process reductions"}
```

Comment:
Is this duplicating the one in connection_coarse_metrics?

Reply:
The labels differ (channels, queues, etc.):

```
# TYPE rabbitmq_erlang_process_reductions counter
# HELP rabbitmq_erlang_process_reductions Erlang process reductions
rabbitmq_erlang_process_reductions_total{channel="<0.937.0>"} 2824434
rabbitmq_erlang_process_reductions_total{channel="<0.973.0>"} 16209
rabbitmq_erlang_process_reductions_total{channel="<0.926.0>"} 2083196
rabbitmq_erlang_process_reductions_total{channel="<0.984.0>"} 11258
rabbitmq_erlang_process_reductions_total{channel="<0.1017.0>"} 17200
# TYPE rabbitmq_erlang_process_reductions counter
# HELP rabbitmq_erlang_process_reductions Erlang process reductions
rabbitmq_erlang_process_reductions_total{channel="<0.976.0>"} 43991
rabbitmq_erlang_process_reductions_total{channel="<0.906.0>"} 1188669
rabbitmq_erlang_process_reductions_total{channel="<0.954.0>"} 49546
rabbitmq_erlang_process_reductions_total{channel="<0.1002.0>"} 48606
rabbitmq_erlang_process_reductions_total{channel="<0.929.0>"} 1993311
rabbitmq_erlang_process_reductions_total{channel="<0.994.0>"} 42323
rabbitmq_erlang_process_reductions_total{channel="<0.947.0>"} 25974
rabbitmq_erlang_process_reductions_total{channel="<0.899.0>"} 25945
# TYPE rabbitmq_erlang_process_reductions counter
# HELP rabbitmq_erlang_process_reductions Erlang process reductions
rabbitmq_erlang_process_reductions_total{vhost="/",queue="basic-get"} 47068
rabbitmq_erlang_process_reductions_total{vhost="/",queue="greedy-consumer"} 6028799
```

Reply:
That's invalid output, you've duplicate metrics there. In addition differing labelnames for a metric is a no-no.

What is the cardinality of these going to be? I'm wondering if this could be one for all connections, one for all queues etc. Otherwise it should be a different metric name each.

Reply:
> That's invalid output, you've duplicate metrics there. In addition differing labelnames for a metric is a no-no.

While we could have more specific metric names, e.g. `rabbitmq_channel_reductions_total` instead of `rabbitmq_erlang_process_reductions_total`, wouldn't this make it harder to answer which Erlang process works the hardest (i.e. has the highest number of reductions)?

What would you recommend instead of this?

```
rabbitmq_erlang_process_reductions_total{vhost="/",queue="basic-get"} 47068
rabbitmq_erlang_process_reductions_total{vhost="/",queue="greedy-consumer"} 6028799
```

> What is the cardinality of these going to be? I'm wondering if this could be one for all connections, one for all queues etc. Otherwise it should be a different metric name each.

The cardinality is going to be high. Would you recommend this as an alternative?

```
rabbitmq_channel_0-976-0_reductions_total 43991
rabbitmq_channel_0-906-0_reductions_total 1188669
rabbitmq_queue_basic-get_reductions_total{vhost="/"} 47068
rabbitmq_queue_greedy-consumer_process_reductions_total{vhost="/"} 6028799
```

Putting IDs such as Erlang process ID in the metric name feels worse than having it as a label, which I agree that is not a good practice, but I am struggling to see an alternative.

Reply:
> The cardinality is going to be high.

Then I'd not break them out by that label, particularly for what I gather is a subtle performance tuning metric.

> Putting IDs such as Erlang process ID in the metric name feels worse than having it as a label

It is a bad idea, what I'm suggesting is not having the process ID at all.

Reply:
If we don't capture the unique identifier for a queue / connection / channel, then we won't have per-object metrics in Prometheus.

What needs to be true for us to answer the following questions using Prometheus metrics?

* Which queue has the most messages?
* Which connection is sending/receiving the most messages?
* Which channel is doing the most work?

Reply:
This is something to be careful with, other systems (e.g. Cassandra, Kafka) are known to be challenging to monitor due to their cardinality for things like this. Scraping at a 5s interval doesn't help.
I'd suggest being quite judicious in what things you break out by these.

Reply:
The scraping interval is now 15s, all metrics with `range()` use 60s consistently (4x).

I understand the challenge, and the fact that there is no good solution that would meet the requirements. Happy to pick this up when we discover that we are putting too much pressure on Prometheus. It works so ship it, but improve long-term.

Reply:
The issue is often too much load on the application, though more than 10k samples from one scrape is where things start to need consideration and 100k is roughly the extreme.

In any case rabbitmq_erlang_process_reductions_total is still resulting in invalid output.

Reply:
The last commit should fix this.

I am happy with this round of improvements, ready to merge and move onto the next one if this is now resolved.